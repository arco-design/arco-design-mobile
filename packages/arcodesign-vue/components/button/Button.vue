<template>
    <button
        :class="[
            prefixCls,
            `${prefixCls}-size-${size}`,
            `${prefixCls}-${type}`,
            `${prefixCls}-shape-${shape}`,
            {
                'is-inline': inline,
                'is-loading': loading,
                'is-disabled': disabled || (disableWhenLoading && loading),
                'is-active': isActive,
            },
            ...customClass,
        ]"
        :style="buttonStyle"
        @touchstart="handleTouchStart"
        @touchend="handleTouchEnd"
        @touchcancel="handleTouchEnd"
        @click="handleClick"
    >
        <span v-if="icon" :class="`${prefixCls}-icon`">
            <slot name="icon">{{ icon }}</slot>
        </span>
        <span v-if="!loading || showTextWhenLoading" :class="`${prefixCls}-text`">
            <slot>{{ children }}</slot>
        </span>
        <!-- Display the loading icon if loading is true -->
        <loading v-if="loading" :class="`${prefixCls}-loading-icon`" />
    </button>
</template>

<script lang="ts">
import { defineComponent, computed, ref, onMounted, watch } from 'vue';
import { ButtonProps, buttonProps } from './type';
import useButtonClass from './hooks';
import Loading from '@/loading';

export default defineComponent({
    name: 'ArcButton',
    components: { Loading },
    props: buttonProps,
    emits: ['click'],
    setup(props: ButtonProps, { emit, slots }) {
        const prefixCls = 'arc-button';
        const isActive = ref(false);

        const handleTouchStart = () => {
            if (props.needActive && !props.disabled) {
                isActive.value = true;
            }
        };

        const handleTouchEnd = () => {
            isActive.value = false;
        };

        const handleClick = (event: Event) => {
            if (props.disabled) {
                props.onClickDisabled?.(event);
                return;
            }
            if (!props.loading || !props.disableWhenLoading) {
                props.onClick?.(event);
            }
        };

        const { buttonClass, buttonStyle } = useButtonClass(props);

        watch(
            () => props.loading,
            newLoading => {
                if (!newLoading) {
                    isActive.value = false;
                }
            },
        );

        onMounted(() => {
            // Handle additional lifecycle logic if needed
        });

        return {
            prefixCls,
            buttonClass,
            buttonStyle,
            isActive,
            handleTouchStart,
            handleTouchEnd,
            handleClick,
        };
    },
});
</script>
