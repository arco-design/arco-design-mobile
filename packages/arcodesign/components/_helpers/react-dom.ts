import { ReactElement } from 'react';
import * as ReactDOM from 'react-dom';

function isObject(obj: any): obj is { [key: string]: any } {
    return Object.prototype.toString.call(obj) === '[object Object]';
}

export interface RootType {
    render: (element: ReactElement) => void;
    _unmount: () => void;
}
export interface RootTypeReact extends RootType {
    unmount?: () => void;
}
export type CreateRootFnType = (container: Element | DocumentFragment) => {
    render: (element: ReactElement) => void;
    unmount?: () => void;
};

// Cast ReactDOM to a version that might have createRoot and other properties
const typedReactDOM = ReactDOM as typeof ReactDOM & {
    createRoot?: CreateRootFnType;
    render?: (
        element: ReactElement,
        container: Element | DocumentFragment,
        callback?: () => void,
    ) => any;
    unmountComponentAtNode?: (container: Element | DocumentFragment) => boolean;
    __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED?: {
        usingClientEntryPoint?: boolean;
    };
};

let copyRender: (
    app: ReactElement,
    container: Element | DocumentFragment,
    createRootFunction?: CreateRootFnType,
) => RootType;

const updateUsingClientEntryPoint = (skipWarning?: boolean) => {
    // https://github.com/facebook/react/blob/17806594cc28284fe195f918e8d77de3516848ec/packages/react-dom/npm/client.js#L10
    // Avoid console warning
    const secretInternals = typedReactDOM.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    if (isObject(secretInternals)) {
        secretInternals.usingClientEntryPoint = skipWarning;
    }
};

const createRootFn = typedReactDOM.createRoot;

const getRender =
    (createRootFunction: CreateRootFnType) =>
    (app: ReactElement, container: Element | DocumentFragment) => {
        updateUsingClientEntryPoint(true);
        const root = createRootFunction!(container);
        updateUsingClientEntryPoint(false);

        root.render(app);

        return {
            render: (elementToRender: ReactElement) => {
                root.render(elementToRender);
            },
            _unmount() {
                setTimeout(() => {
                    if (root && typeof root.unmount === 'function') {
                        root.unmount();
                    }
                });
            },
        };
    };

if (createRootFn) {
    // React 18
    copyRender = getRender(createRootFn);
} else if (
    typeof typedReactDOM.render === 'function' &&
    typeof typedReactDOM.unmountComponentAtNode === 'function'
) {
    // React 16/17
    copyRender = function (app: ReactElement, container: Element | DocumentFragment) {
        typedReactDOM.render!(app, container); // Use non-null assertion
        return {
            render: (elementToRender: ReactElement) => {
                typedReactDOM.render!(elementToRender, container); // Use non-null assertion
            },
            _unmount() {
                typedReactDOM.unmountComponentAtNode!(container); // Use non-null assertion
            },
        };
    };
} else {
    copyRender = (
        app: ReactElement,
        container: Element | DocumentFragment,
        createRootFunction?: CreateRootFnType,
    ) => {
        const defaultCb = () => {
            // Fallback if no rendering method is found
            console.error(
                'ArcoDesign: ReactDOM.createRoot() or ReactDOM.render() not found. ' +
                    'This usually means you are using an unsupported version of React, ' +
                    'or ReactDOM is not properly initialized. ' +
                    'ArcoDesign Mobile React requires React 16, 17, 18, or 19.',
            );
            return { render: (_element: ReactElement) => {}, _unmount: () => {} };
        };
        if (createRootFunction) {
            return getRender(createRootFunction)(app, container);
        }
        return defaultCb();
    };
}

export const render = copyRender;
