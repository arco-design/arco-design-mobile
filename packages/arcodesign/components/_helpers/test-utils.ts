import React, { useEffect, useRef } from 'react';
import { Promise as ES6Promise, Promise } from 'es6-promise';

/**
 * React 19 兼容的 act 包装器
 * 用于包装可能触发状态更新的异步操作
 */
export const actWrapper = (callback: () => void | ES6Promise<void>): ES6Promise<void> => {
    return new ES6Promise<void>((resolve, reject) => {
        try {
            // 尝试使用 react-dom/test-utils 的 act
            const testUtils = require('react-dom/test-utils');
            if (testUtils && typeof testUtils.act === 'function') {
                // 对于同步回调
                if (typeof callback === 'function') {
                    try {
                        const result = callback();
                        if (result && typeof result.then === 'function') {
                            testUtils
                                .act(() => result)
                                .then(() => resolve())
                                .catch(reject);
                        } else {
                            testUtils.act(() => {
                                callback();
                            });
                            resolve();
                        }
                    } catch (error) {
                        reject(error);
                    }
                } else {
                    resolve();
                }
            } else {
                const result = callback();
                if (result && typeof result.then === 'function') {
                    result.then(() => resolve()).catch(reject);
                } else {
                    resolve();
                }
            }
        } catch (error) {
            // 如果 act 不可用，直接执行回调
            try {
                const result = callback();
                if (result && typeof result.then === 'function') {
                    result.then(() => resolve()).catch(reject);
                } else {
                    resolve();
                }
            } catch (callbackError) {
                reject(callbackError);
            }
        }
    });
};

/**
 * 清理未处理的异步操作，防止内存泄漏警告
 */
export const cleanupAsync = (): ES6Promise<void> => {
    return new ES6Promise<void>(resolve => {
        setTimeout(() => resolve(), 0);
    });
};

/**
 * 安全的状态更新，检查组件是否已卸载
 */
export const createSafeStateUpdater = (isMountedRef: React.MutableRefObject<boolean>) => {
    return (updater: () => void) => {
        if (isMountedRef.current) {
            updater();
        }
    };
};

/**
 * 用于测试的 useEffect 清理函数
 */
export const useTestCleanup = (cleanupFn: () => void) => {
    const isMountedRef = useRef(true);

    useEffect(() => {
        return () => {
            isMountedRef.current = false;
            cleanupFn();
        };
    }, [cleanupFn]);

    return isMountedRef;
};
