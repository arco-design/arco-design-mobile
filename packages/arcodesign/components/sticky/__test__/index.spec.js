import React from 'react';
import demoTest from '../../../tests/demoTest';
import mountTest from '../../../tests/mountTest';
import Sticky from '..';
import Button from '../../button';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { mockAddListener } from '../../../tests/helpers/mockEvent';

demoTest('sticky');

mountTest(Sticky, 'Sticky');

describe('Sticky', () => {

    beforeEach(() => {
        document.body.getBoundingClientRect = jest.fn(() => ({
            top: 100,
            bottom: 300
        }))
        jest.spyOn(window, 'requestAnimationFrame').mockImplementation(cb => cb());
        jest.useFakeTimers();
    });

    afterEach(() => {
        window.requestAnimationFrame.mockRestore();
        jest.useRealTimers();
    });

    it('should callback correctly when scrolled', () => {
        const mockFn = jest.fn();
        const component = mount(
            <Sticky topOffset={64} onTopChange={mockFn}>
                <Button>Test</Button>
            </Sticky>
        );
        const map = mockAddListener(window, true);
        component.setProps({ topOffset: 0 });
        act(() => {
            map.scroll({currentTarget: window});
        });
        expect(mockFn).toBeCalled();
    });
});
