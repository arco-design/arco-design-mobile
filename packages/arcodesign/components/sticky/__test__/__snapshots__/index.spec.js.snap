// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sticky demo test sticky demo: bottom.md renders correctly 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div
        style="height: 0px;"
      />
      <div
        class="arco-sticky"
      >
        <button
          class="arco-button arco-button-type-primary type-primary arco-button-size-large arco-button-size-large-is-semi size-large is-semi pc"
          type="button"
        >
          <div
            class="arco-button-text arco-button-text-pc btn-text"
          >
            Fixed bottom, global scrolling, 40px position relative to the bottom
          </div>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`sticky demo test sticky demo: index.md renders correctly 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div
        style="height: 0px;"
      />
      <div
        class="arco-sticky"
      >
        <button
          class="arco-button arco-button-type-primary type-primary arco-button-size-large arco-button-size-large-is-semi size-large is-semi pc"
          type="button"
        >
          <div
            class="arco-button-text arco-button-text-pc btn-text"
          >
            Global scrolling, positioned relative to the top 64px
          </div>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`sticky demo test sticky demo: offset.md renders correctly 1`] = `
<DocumentFragment>
  <div
    id="container"
  >
    <div>
      <div
        style="height: 0px;"
      />
      <div
        class="arco-sticky"
      >
        <button
          class="arco-button arco-button-type-primary type-primary arco-button-size-large arco-button-size-large-is-semi size-large is-semi pc"
          type="button"
        >
          <div
            class="arco-button-text arco-button-text-pc btn-text"
          >
            100px to the top, follows when the container leaves the viewport
          </div>
        </button>
      </div>
    </div>
    <div
      class="demo-sticky-container"
    >
      Bottom of container
    </div>
  </div>
</DocumentFragment>
`;

exports[`sticky demo test sticky demo: relative.md renders correctly 1`] = `
<DocumentFragment>
  <div
    style="height: 300px; overflow: scroll;"
  >
    <div
      class="demo-sticky-container"
      style="height: 200px;"
    >
      My normal content
    </div>
    <div>
      <div
        style="height: 0px;"
      />
      <div
        class="arco-sticky"
      >
        <button
          class="arco-button arco-button-type-primary type-primary arco-button-size-large arco-button-size-large-is-semi size-large is-semi pc"
          type="button"
        >
          <div
            class="arco-button-text arco-button-text-pc btn-text"
          >
            Relative scroll content
          </div>
        </button>
      </div>
    </div>
    <div
      class="demo-sticky-container"
      style="height: 500px;"
    >
      My normal content
    </div>
  </div>
</DocumentFragment>
`;
